src/sunmatrix/dense/sunmatrix_dense.c

```diff
@@ -286,70 +297,66 @@ int SUNMatMatvec_Dense(SUNMatrix A, N_Vector x, N_Vector y)
   sunindextype i, j;
   realtype *col_j, *xd, *yd;
 
-  /* Verify that A, x and y are compatible */
-  if (!SMCompatible2_Dense(A, x, y))
+  if (!compatibleMatrixAndVectors(A, x, y))
     return SUNMAT_ILL_INPUT;
 
-  /* access vector data (return if failure) */
+  /* access vector data (return if NULL data pointers) */
   xd = N_VGetArrayPointer(x);
   yd = N_VGetArrayPointer(y);
   if ((xd == NULL) || (yd == NULL) || (xd == yd))
     return SUNMAT_MEM_FAIL;
 
-  /* Perform operation */
-  for (i=0; i<SM_ROWS_D(A); i++)
+  /* Perform operation y = Ax */
+  for (i = 0; i < SM_ROWS_D(A); i++)
     yd[i] = ZERO;
-  for(j=0; j<SM_COLUMNS_D(A); j++) {
-    col_j = SM_COLUMN_D(A,j);
-    for (i=0; i<SM_ROWS_D(A); i++)
-      yd[i] += col_j[i]*xd[j];
+  for (j = 0; j < SM_COLUMNS_D(A); j++) {
+    col_j = SM_COLUMN_D(A, j);
+    for (i = 0; i < SM_ROWS_D(A); i++)
+      yd[i] += col_j[i] * xd[j];
   }
+
   return SUNMAT_SUCCESS;
 }
 
-int SUNMatSpace_Dense(SUNMatrix A, long int *lenrw, long int *leniw)
+int SUNMatSpace_Dense(SUNMatrix A, long int* lenrw, long int* leniw)
 {
   *lenrw = SM_LDATA_D(A);
   *leniw = 3 + SM_COLUMNS_D(A);
   return SUNMAT_SUCCESS;
 }
 
-
 /*
  * -----------------------------------------------------------------
  * private functions
  * -----------------------------------------------------------------
  */
 
-static booleantype SMCompatible_Dense(SUNMatrix A, SUNMatrix B)
+static booleantype compatibleMatrices(SUNMatrix A, SUNMatrix B)
 {
   /* both matrices must be SUNMATRIX_DENSE */
-  if (SUNMatGetID(A) != SUNMATRIX_DENSE)
-    return SUNFALSE;
-  if (SUNMatGetID(B) != SUNMATRIX_DENSE)
+  if ((SUNMatGetID(A) != SUNMATRIX_DENSE) || (SUNMatGetID(B) != SUNMATRIX_DENSE)) {
     return SUNFALSE;
+  }
 
   /* both matrices must have the same shape */
-  if (SM_ROWS_D(A) != SM_ROWS_D(B))
-    return SUNFALSE;
-  if (SM_COLUMNS_D(A) != SM_COLUMNS_D(B))
+  if ((SM_ROWS_D(A) != SM_ROWS_D(B)) || (SM_COLUMNS_D(A) != SM_COLUMNS_D(B))) {
     return SUNFALSE;
+  }
 
   return SUNTRUE;
 }
 
-
-static booleantype SMCompatible2_Dense(SUNMatrix A, N_Vector x, N_Vector y)
+static booleantype compatibleMatrixAndVectors(SUNMatrix A, N_Vector x, N_Vector y)
 {
-  /*   vectors must be one of {SERIAL, OPENMP, PTHREADS} */
-  if ( (N_VGetVectorID(x) != SUNDIALS_NVEC_SERIAL) &&
-       (N_VGetVectorID(x) != SUNDIALS_NVEC_OPENMP) &&
-       (N_VGetVectorID(x) != SUNDIALS_NVEC_PTHREADS) )
+  /* Vectors must provide nvgetarraypointer and cannot be a parallel vector */
```

Comment:
Is it not having a communicator really a necessity?  

If we removed this stipulation then for a rank-decoupled linear system, it would be straightforward for each MPI rank to apply a rank-local dense/banded/sparse linear solver on their local portion of an MPI-parallel vector (assuming the dimensions line up).  I'm specifically thinking of an ImEx or Multirate problem where implicitness is relegated to rank-local operators.  We've previously discussed being a little more "trusting" of our users (with big warning labels in the documentation), so removing the `N_VGetCommunicator` checks here would be a step in that direction.

Reply:
I would be for relaxing the check and agree with you reasoning. I think we discussed this maybe didn't reach a final conclusion (or I don't remember).

Reply:
@gardner48 Thoughts?

Reply:
I think this sounds reasonable. With the error handling updates, we could have a warning about possible incompatibility only for debug builds and not release builds.